existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,264 @@
+
+```haskell
+example :: a -> [a] -> [a]
+example x rest = double ++ rest
+  where
+    double :: a -> [a]
+    double x = [x, x]
+```
+This will cause an error without `ScopedTypeVariables`. But to use it, you need to add a `forall.` quantifier.
+
+```haskell
+example :: forall a. a -> [a] -> [a]
+```
+You might also notice how it was useful with the `TypeApplication` extension.
+Other extensions that somehow benefit from the usage of `ExplicitForAll` are `LiberalTypeSynonyms`, `RankNTypes`, and many more.
+
+## Existential quantification
+Besides universally quantified variables, Haskell also supports existential types.
+But in order to use them ... you still need to use the `forall.` keyword ¯\\\_(ツ)_/¯.
+
+The following construction is called an existential type in Haskell – when a type variable is introduced (using the `forall.` quantifier) inside the signature, not before it.
+In contrast to the universal quantifier, the existential one means that the compiler (**callee**) is responsible for deducing what a type variable should mean.
+This also means that the caller can't specify this type variable.
+
+![When you use existential types](meme.jpg)
+
+
+### Usage
+
+We will start with functions:
+
+```haskell
+func :: forall a c. a -> (forall b. (Show b) => b) -> c
+```
+There are a few things to mention here.
+
+* In the type signature, `a` and `c` are universally quantified variables, and `b` is existentially quantified. This is what I was talking about when I said that type variables may belong to different quantifiers.
+* The same variable can't be quantified more than once.
+* You can see that constraints for `b` are added in the middle of the declaration of the function type (because `b` was introduced in the middle of the function).
+* Now you can't manually choose type for `b` – the compiler must deduce the correct variable you want to use by itself.
+
+Here's a real-life example of using existential types in functions.
+Imagine a function that prints some logs while calculating the result.
+You may want to write logs into a console or into some file.
+In order to do it, you pass a logging function as an argument.
+
+```haskell
+func :: _ -> IO ()
+func log = do
+  username <- getUsername
+  log username
+
+  userCount <- getUserCount
+  log userCount
+
+getUsername :: IO String
+getUsername = undefined
+
+getUserCount :: IO Int
+getUserCount = undefined
+```
+
+What type should the log function have?
+Since it somehow logs both `String` and `Int`:
+
+```haskell
+log :: Show a => a -> IO ()
+```
+However, if we add this signature to `func` and try to call it with different arguments:
+
+```haskell
+func :: Show a => (a -> IO ()) -> IO ()
+func log = ...
+
+main :: IO ()
+main = do
+  -- log to the console
+  func print
+
+  -- log to a logfile
+  func (appendFile "logfile.log" . show)
+```
+We will see some nasty compiler errors:
+```
+Ambiguous type variable ‘a1’ arising from a use of ‘func’
```

Comment:
The way this is phrased, it appears that the problem with the code snippet above is in `main`.

But it's not - the problem is in `func`. `func`'s definition is trying to instantiate `a=String`, but `a` is universally quantified, and that's why it can't be compiled. `func`'s definition cannot instantiate `a`, only the caller of `func` can.

![image](https://user-images.githubusercontent.com/718166/152653881-72475ac8-0463-4d6e-8be2-2d073588ce1e.png)

Yes, there's a "Ambiguous type variable" error in `main`, but that's a red herring. The real cause of the issue is in `func`.

Reply:
👏 

@dcastro your comments are the reason why expert reviews are as important, if not more important as "beginners' mind" reviews. Amazing, stellar work! 

I want to say that I don't want to diminish @PruStephan's work too. Writing blogs are hard and for the first attempt, this is a very very good start!

Reply:
Much appreciated! :smile: :bow: 

> Writing blogs are hard and for the first attempt, this is a very very good start!

Indeed! I want to echo your words.

Writing blog posts is also a great learning & discovery process.
@PruStephan I can tell you've learned a lot since you began writing this! Which can only mean you're doing a good job :smile:

Reply:
Thanks. That is actually was the reason I am writing this post, to learn something new. I think I after this post, I will continue with writing